ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -277,34 +351,424 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
+		size_t high_i = i + half_size;
+		n1_buf[i] = n1[high_i] - n1[i];
+		n2_buf[i] = n2[high_i] - n2[i];
+	}
+
+	/* mid */
+	BC_UINT_T *mid = high + calc_size;
+	bc_rec_mul_recursive_fast(n1_buf, n1_buf + half_size, n2_buf, n2_buf + half_size, half_size, mid);
+
+	/*
+	 * Add to ret.
+	 * From 0 to half_size, adding it to ret will affect the low value, so add it later.
+	 */
+	for (i = 0; i < half_size; i++) {
+		mid[i] -= low[i] + high[i];
+	}
+	for (; i < calc_size; i++) {
+		ret[i + half_size] += low[i] + high[i] - mid[i];
+	}
+	for (i = 0; i < half_size; i++) {
+		ret[i + half_size] -= mid[i];
+	}
+
+	/*
+	 * Perform digit adjustment to avoid overflow.
+	 */
+	if (UNEXPECTED(bc_rec_mul_near_overflow(calc_size))) {
+		for (i = 0; i < calc_size * 2; i++) {
+			bc_digits_adjustment_single(ret + i);
+		}
+	}
+}
+
+/*
+ * This is a function that is called recursively in divide-and-conquer calculations.
+ * Compute the three multiplications of Karatsuba-algorithm as high, mid, and low, respectively.
+ * e.g. 1234 * 5678
+ * low = 34 * 78
+ * high = 12 * 56
+ * mid = (12 - 34) * (56 - 78)
+ * ret = high * 100^2 + (high + low - mid) * 100 + low
+ * This function differs from "fast" in that it does things like check the number of digits.
+ */
+static void bc_rec_mul_recursive(
+	const BC_UINT_T *n1, BC_UINT_T *n1_buf, size_t n1_size,
+	const BC_UINT_T *n2, BC_UINT_T *n2_buf, size_t n2_size,
+	size_t calc_size, BC_UINT_T *ret)
+{
+	size_t i;
+	size_t half_size = calc_size / 2;
+
+	/*
+	 * Find the size of the values ​​used to calculate high and low. It also checks whether
+	 * to skip calculating high.
+	 */
+	bool fast;
+	bool skip_high;
+	size_t n1_low_size;
+	size_t n1_high_size;
+	size_t n2_low_size;
+	size_t n2_high_size;
+	if (n1_size > half_size) {
+		n1_low_size = half_size;
+		n1_high_size = n1_size - half_size;
+	} else {
+		n1_low_size = n1_size;
+		n1_high_size = 0;
+	}
+	if (n2_size > half_size) {
+		n2_low_size = half_size;
+		n2_high_size = n2_size - half_size;
+		skip_high = false;
+		fast = true;
+	} else {
+		n2_low_size = n2_size;
+		n2_high_size = 0;
+		skip_high = true;
+		fast = n2_size == half_size;
+	}
+
+	/* low */
+	BC_UINT_T *low = ret;
+	size_t low_ret_size = n1_low_size + n2_low_size;
+	if (fast) {
+		bc_rec_mul_recursive_fast(n1, n1_buf, n2, n2_buf, half_size, low);
+	} else {
+		bc_rec_mul_recursive(n1, n1_buf, n1_low_size, n2, n2_buf, n2_low_size, half_size, low);
+	}
+
+	/* high */
+	BC_UINT_T *high = low + low_ret_size;
+	size_t high_ret_size = n1_high_size + n2_high_size;
+	if (!skip_high) {
+		bc_rec_mul_recursive(n1 + half_size, n1_buf, n1_high_size, n2 + half_size, n2_buf, n2_high_size, half_size, high);
+	} else {
+		/*
+		 * Actually, should think of filling calc_size - low_ret_size with 0 and filling
+		 * high_ret_size - (calc_size - low_ret_size) with 0, but the result is the same,
+		 * so write it like this.
+		 *
+		 * Also, in the first called bc_rec_mul, this zero padding may slightly overrun
```

Comment:
I didn't understand this

Reply:
It is mistake. This doesn't actually overrun. I'm going to delete this comment.